app/src/androidTest/java/fr/free/nrw/commons/UploadTest.java

```diff
@@ -0,0 +1,131 @@
+            throw new Error("This test should only be run in Beta!");
+        }
+
+        // Uri to return by our mock gallery selector
+        // Requires file 'image.jpg' to be placed at root of file structure
+        Uri imageUri = Uri.parse("file://mnt/sdcard/image.jpg");
+
+        // Build a result to return from the Camera app
+        Intent intent = new Intent();
+        intent.setData(imageUri);
+        ActivityResult result = new ActivityResult(Activity.RESULT_OK, intent);
+
+        // Stub out the File picker. When an intent is sent to the File picker, this tells
+        // Espresso to respond with the ActivityResult we just created
+        intending(allOf(hasAction(Intent.ACTION_GET_CONTENT), hasType("image/*"))).respondWith(result);
+
+        // Open FAB
+        onView(allOf(withId(R.id.fab_plus), isDisplayed()))
+                .perform(click());
+
+        // Click gallery
+        onView(allOf(withId(R.id.fab_gallery), isDisplayed()))
+                .perform(click());
+
+        // Validate that an intent to get an image is sent
+        intended(allOf(hasAction(Intent.ACTION_GET_CONTENT), hasType("image/*")));
+
+        // Create filename with the current time (to prevent overwrites)
+        SimpleDateFormat dateFormat = new SimpleDateFormat("yyMMdd-hhmmss");
+        String commonsFileName = "MobileTest " + dateFormat.format(new Date());
+
+        // Try to dismiss the error, if there is one (probably about duplicate files on Commons)
+        try {
+            onView(withText("Yes"))
+                    .check(matches(isDisplayed()))
+                    .perform(click());
+        } catch (NoMatchingViewException ignored) {}
+
+        onView(allOf(withId(R.id.description_item_edit_text), withParent(withParent(withId(R.id.image_title_container)))))
+                .perform(replaceText(commonsFileName));
+
+        onView(withId(R.id.bottom_card_next))
+                .perform(click());
+
+        try {
+            Thread.sleep(500);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+
+        onView(withId(R.id.category_search))
+                .perform(replaceText("Uploaded with Mobile/Android Tests"));
+
+        try {
+            Thread.sleep(3000);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+
+        onView(withParent(withId(R.id.categories)))
+                .perform(click());
+
+        onView(withId(R.id.category_next))
+                .perform(click());
+
+        try {
+            Thread.sleep(500);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+
+        onView(withId(R.id.submit))
+                .perform(click());
+
+        try {
+            Thread.sleep(10000);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+
+        String fileUrl = "https://commons.wikimedia.beta.wmflabs.org/wiki/File:" +
```

Comment:
Can we enhance this bit as well? We can probably call media wiki api to assert if the image was uploaded.